guest/file_manager.go

```diff
@@ -127,7 +127,7 @@ func (m FileManager) TransferURL(ctx context.Context, u string) (*url.URL, error
 		return nil, err
 	}
 
-	if !m.c.IsVC() {
+	if url.Hostname() != "*" {
```

Comment:
The call above (`m.c.ParseURL`) already replaces `*` with the connection hostname: https://github.com/vmware/govmomi/blob/4316838a111c783776b1d9e06a6f7f61e507feeb/vim25/soap/client.go#L600-L602

So `url.Hostname()` would never be == `*` at this point, right?  Is your initial client connection directly to an ESX host or to vCenter?

Reply:
Indeed, my proposal is incorrect and `Hostname` should never be == `*`.

The current govc client behavior is, in case it's connected to vCenter, it replaces the url's hostname part by the ip of the ESX host.
In our setup, the ESX hosts have private IP making the `vc.Client.Upload` call to fail.
Our vCenter returns an url that shouldn't be modified by the client and be used directly as it is to access the ESX host.

To not break dependent packages and follow the specs, shouldn't the client try first the url as it is returned by the vCenter and modify the hostname only if the first attempt fails ?

Reply:
ok, how about instead of calling `m.c.ParseURL()`, just call `url.Parse()` directly in place, something like:
```go
        needsHostname := turl.Hostname() = "*"
        if needsHostname {
                turl.Host = u.Host
        }
	if !m.c.IsVC() {
		return turl, nil // we already connected to the ESX host and have its thumbprint
	}
```
(note: renamed the `url` var to `turl` so the url package can be imported)

Then add an `if needsHostname {}` around this block of code, so we still collect the thumbprint without overwriting the hostname: https://github.com/vmware/govmomi/blob/4316838a111c783776b1d9e06a6f7f61e507feeb/guest/file_manager.go#L166-L183

Reply:
The go client is connected to vSphere via public internet.
The ESX host are in a private network and as a client, I can't access that network directly. 

The content of the InitiateFileTransferToGuestResponse from vSphere is
```xml
	<soapenv:Body>
		<InitiateFileTransferToGuestResponse xmlns="urn:vim25">
			<returnval>https://pcc-aaa-bbb-ccc-ddd.ovh.com:40050/guestFile?id=23&amp;token=xxx</returnval>
		</InitiateFileTransferToGuestResponse>
	</soapenv:Body>
```

My understanding is that because the client is connected to vSphere, `m.c.IsVC` will evaluate to `true` and will replace the hostname part of the url by the ESX host private IP. Later the `Upload` \ `Download` calls will fail

I don't expect that behavior.

I'm expecting the client to trust and follow the urls returned in the `InitiateFileTransferToGuestResponse` / `InitiateFileTransferFromGuestResponse`, without modification.

Is that possible ?

Reply:
Yes, with the suggestions I made above, adding `if needsHostname {}` above avoids the hostname from being overwritten when `InitiateFileTransferToGuestResponse.returnval != "*"`.  And preserves the original behavior of overwriting when it is `== *`.  And with the suggested changes, the host thumbprint is collected in both cases, rather than just returning early.